### ISSUE TITLE
Can't override actions.add_new_branch

### ISSUE THREAD

**Issue by vunhatchuong** [xavierchanth/arbor.nvim#2]
```lua
  {
      "xavierchanth/arbor.nvim",
      cmd = { "Arbor" },
      ---@type arbor.config
      opts = {
          apply_recommended = false,
          worktree = {
              normal = { base = "relative_common", path = "../.." },
              bare = { base = "relative_common", path = "../.." },
          },
          settings = {
              add = {
                  on_existing = Util.worktree.arbor_post_switch,
                  path_style = "smart",
              },
          },
          actions = {
              add = {
                  ["add new branch"] = function(info)
                      require("arbor").actions.add_new_branch(
                          info,
                          { path_style = "smart" }
                      )
                  end,
              },
              pick = {
                  ["remove worktree"] = function()
                      require("arbor").remove()
                  end,
              },
          },
          hooks = {
              post_add = function(info)
                  return Util.worktree.arbor_post_add(info)
              end,
              post_pick = function(info)
                  return Util.worktree.arbor_post_switch(info)
              end,
              pre_remove = function(info)
                  return Util.worktree.arbor_pre_remove(info)
              end,
          },
      },
  }
```

Here I override the  `add new branch`  action with `{ path_style = "smart" }` but it still prompts for input.

**Comment by vunhatchuong** [comment]
Also, it seems the settings of `opts.settings.add` don't apply globally? When I set `path_style = "smart"` in `settings.add` I shouldn't need to set it anywhere else.

**Comment by xavierchanth** [comment]
> Here I override the  `add new branch`  action with `{ path_style = "smart" }` but it still prompts for input.

Ah, so when you use the add picker if it is a remote branch it will use "smart" to create the branch.

But if it is a local branch, it will call the on_existing function which should, by default (I think), trigger a switch since the branch already exists. 

What add new branch does, is it forcefully overrides path_style to prompt, so that you can duplicate a local branch. Maybe I can make it so that by default the add picker will ignore the on_existing flow if the function is nil, and set the default to nil? This would remove the need to use "add new branch" in most scenarios.

But, even if the path_style is smart, it would fallback to prompt when smart is not able to resolve a path name automatically. This is true when a path or branch name can't be resolved.



**TLDR**: I'll try to make things more natural around add, and drop the need for the "add new branch" action.

Add has a lot of edge cases, and I tried to avoid the need to make too many git queries, but I may need to make a few more to get things just right. So far I've not seen performance issues, but I have a feeling they will manifest more on Windows...

**Comment by vunhatchuong** [comment]
> But if it is a local branch, it will call the on_existing function which should, by default (I think), trigger a switch since the branch already exists.

Are you referring to the default "add" action or the custom "add new branch"? To clarify, the bug I mentioned occurs when I use the custom "add new branch."  

The issue came up when I tried creating a folder named `haha` for a new branch called `fix/nvim-error`, based on `main`. It keeps placing the folder inside `./fix/nvim-error` instead of creating it where I expected. 

---

Here’s the flow I have in mind:  

1. While in folder A on the `main` branch, I use `:Arbor add`.  
2. This prompts the user to provide a name for the new branch and a folder.  
3. If the user selects an existing branch, it still creates a new folder with the prompted name. The folder's location is determined by the settings.  

I like the idea of having a single command handle both adding and switching, but there should still be a way to add a branch even if it already exists.

**Comment by xavierchanth** [comment]
> > But if it is a local branch, it will call the on_existing function which should, by default (I think), trigger a switch since the branch already exists.
> 
> 
> 
> Are you referring to the default "add" action or the custom "add new branch"? To clarify, the bug I mentioned occurs when I use the custom "add new branch."  
> 
> 
> 
> The issue came up when I tried creating a folder named `haha` for a new branch called `fix/nvim-error`, based on `main`. It keeps placing the folder inside `./fix/nvim-error` instead of creating it where I expected. 
> 
> 
> 
> ---
> 
> 
> 
> Here’s the flow I have in mind:  
> 
> 
> 
> 1. While in folder A on the `main` branch, I use `:Arbor add`.  
> 
> 2. This prompts the user to provide a name for the new branch and a folder.  
> 
> 3. If the user selects an existing branch, it still creates a new folder with the prompted name. The folder's location is determined by the settings.  
> 
> 
> 
> I like the idea of having a single command handle both adding and switching, but there should still be a way to add a branch even if it already exists.

Oh, I misunderstood the original report. That is most definitely a bug, most likely a path which should be absolute is relative. I'll see if I can take a look after my work day.

**Comment by xavierchanth** [comment]
Hey @vunhatchuong, I'm not able to reproduce, would you be able to provide logs from the pre_add and post_add hooks:

```
pre_add = function(info)
  vim.print("pre")
  vim.print(vim.inspect(info))
end,
post_add = function(info)
    vim.print("post")
  vim.print(vim.inspect(info))
  return Util.worktree.arbor_post_add(info)
end,
```

Thanks in advance

**Comment by vunhatchuong** [comment]
The original setup only has 1 branch `main` both locally and upstream.

Steps:

1. `:Arbor add`.
2. Choose `action: add new branch`.
3. Choose `branch: main`.
4. Type "Path for the worktree" with `haha`.


log:
```bash
post
{
  branch = "main",
  branch_info = {
    display_name = "main",
    head = true,
    refname = "refs/heads/main",
    upstream = "origin/main",
    worktree_path = "/home/ronny/.dotfiles"
  },
  common_dir = "/home/ronny/.dotfiles/.git",
  cwd = "/home/ronny/.dotfiles/nvim",
  new_branch = "haha",
  new_path = "/home/ronny/haha",
  operation_opts = {
    branch_style = "path",
    hooks = {
      pre = <function 1>
    },
    on_existing = false,
    path_style = "prompt",
    show_actions = false,
    show_remote_branches = true
  },
  repo_type = "normal",
  resolved_base = "/home/ronny",
  toplevel = "/home/ronny/.dotfiles"
}
````

**Comment by xavierchanth** [comment]
So the expect path was `/home/ronny/haha`, but the worktree was created under `/home/ronny/.dotfiles/nvim/haha`?

**Comment by vunhatchuong** [comment]
No, the expected result is `path_style = "smart"` not `path_style = "prompt"` thus it **shouldn't prompt** me to input the path at all.

Although unrelated to this issue, I expect `base = "relative_common"` to put me at the root of repo and not inside `.git` folder so I can just use `path = "../"` and not `path = "../.."` like above.

**Comment by xavierchanth** [comment]
> No, the expected result is `path_style = "smart"` not `path_style = "prompt"` thus it **shouldn't prompt** me to input the path at all.

How would you expect `smart` to come up with a new name from nothing? `smart` will name your local branch the same as the remote branch when adding from the remote with the base add picker. But when a local branch already exists and you are trying to create a new branch and worktree, I have no way of determining a "smart" name. I can't name them the same as something that already exists, so the add new branch will fallback to a prompt to resolve the non-deterministic information. Note that path is always resolved first, but by default this will set the branch name to the path name.

If you have a particular naming scheme, there is already a function for specifying your own custom path resolver, but the add new branch will still override to prompt. I can modify that action to only override the builtin resolvers like "smart" and "basename"

**Comment by xavierchanth** [comment]
I patched the add action so you can specify a custom resolver function.
You can find examples of the builtin ones here: https://github.com/XavierChanth/arbor.nvim/blob/326a7b915ad71763ac4486476ad2bd45254cb57c/lua/arbor/core/add.lua#L91-L121

```lua
local function custom_path_resolver(info, local_branches)
---info type spec (dictionary table):
---@class arbor.git.info
---@field operation_opts? arbor.opts
---@field branch string
---@field cwd string
---@field toplevel string
---@field repo_type arbor.git.repo_type
---@field common_dir string
---@field resolved_base? string
---@field branch_info? arbor.git.branch
---@field new_path? string
---@field new_branch? string


---local_branches (list of tables)
---@class arbor.git.branch
---@field head boolean
---@field main_branch boolean?
---@field refname string
---@field upstream string?
---@field worktree_path string?
---@field display_name string
end
-- In your opts:
path_style = custom_path_resolver
```

Note that some values in info will be null at this point, specifically, new_path and new_branch, since this is used to resolve them.

**Comment by xavierchanth** [comment]
> Although unrelated to this issue, I expect `base = "relative_common"` to put me at the root of repo and not inside `.git` folder so I can just use `path = "../"` and not `path = "../.."` like above.

In git, for a given repo the only fixed path we have which can be resolved is the git common dir, and this is not always in the same place in every repository structure. Specifically in a default bare repo it is already the "root".

I can also add a relative_workdir which will resolve to the current git working directory, but I haven't as of yet, because people working with nested paths will experience unusual behavior.

**Comment by vunhatchuong** [comment]
> How would you expect smart to come up with a new name from nothing ...

Sorry I was too tunneled trying to make my workflow work, but I think found the problem.

https://github.com/XavierChanth/arbor.nvim/blob/326a7b915ad71763ac4486476ad2bd45254cb57c/lua/arbor/actions/add_new_branch.lua#L22

should be `opts.path_style = opts.path_style or "prompt"` otherwise it will override the user's settings.

---

Also, in this setting `branch_style = "prompt"` is overridden correctly but `path_style` is not, it still prompts for input instead of taking the basename.

```lua
opts = {
    settings = {
        add = {
            path_style = "basename",
            branch_style = "prompt",
        },
    }}
```

If I put it directly into "add new branch" it does override correctly

```lua
actions = {
    add = {
        ["add new branch"] = function(info)
            require("arbor").actions.add_new_branch(info, {
                path_style = "basename",
                branch_style = "prompt",
            })
        end,
    }}
```

---

Also, I guess this is personal preference but I expect `path_style = "basename"` to take the basename of the newly created branch not the name of the branch that it inherits from.

So If I create a new branch named `fix/nvim-error` from `main` the `basename` should be `nvim-error` and not `main`.

**Comment by xavierchanth** [comment]
> should be `opts.path_style = opts.path_style or "prompt"` otherwise it will override the user's settings.

The override was intentional, but I've loosened it to accept your own custom function in the latest release. See this comment:

https://github.com/XavierChanth/arbor.nvim/issues/2#issuecomment-2488596067

Regardless of whether the path_style is smart or basename, still can't create something from nothing. If you have some clever way you want to resolve names, use the function, but I won't be adding that to the core. Even if I come up with something decent and not prone to bugs, it's likely to be opinionated and a source for complaints.

**Comment by xavierchanth** [comment]
> Also, I guess this is personal preference but I expect `path_style = "basename"` to take the basename of the newly created branch not the name of the branch that it inherits from.
> 
> So If I create a new branch named `fix/nvim-error` from `main` the `basename` should be `nvim-error` and not `main`.

Nope, that one is a bug... I changed some things around early on in development and that is referencing the wrong path, I'll have a release with a fix.

**Comment by xavierchanth** [comment]
> > Also, I guess this is personal preference but I expect `path_style = "basename"` to take the basename of the newly created branch not the name of the branch that it inherits from.
> > So If I create a new branch named `fix/nvim-error` from `main` the `basename` should be `nvim-error` and not `main`.
> 
> Nope, that one is a bug... I changed some things around early on in development and that is referencing the wrong path, I'll have a release with a fix.

Actually, no it's not. Same story there as above. What you are looking for is `branch_style="path"`, which will call the branch `fix/nvim-error`

**Comment by xavierchanth** [comment]
Note that path is always resolved before branch, so it is the branch name that is determined based on the path name, not the other way around. Branch has a dependency on path, and this is mainly for the reason that branch name can be unset (`branch_style=git` will tell git worktree to **try** to infer a branch name, note the emphasis on try). So it's not feasible to trust path having a dependency on branch.

**Comment by xavierchanth** [comment]
You can assume with the defaults that when you set path, you are setting both branch and path.

**Comment by vunhatchuong** [comment]
> Note that path is always resolved before branch

Oh I see, so I can't do something like this also:

```lua
actions = {
    add = {
        ["add new branch"] = function(info)
            require("arbor").actions.add_new_branch(info, {
                path_style = function()
                    return vim.fs.basename(info.new_branch)
                end,
                branch_style = "prompt",
            })
        end}}
```

Welp... Sorry to bother you for nothing then.

**Comment by xavierchanth** [comment]
> > Note that path is always resolved before branch
> 
> 
> 
> Oh I see, so I can't do something like this also:
> 
> 
> 
> ```lua
> 
> actions = {
> 
>     add = {
> 
>         ["add new branch"] = function(info)
> 
>             require("arbor").actions.add_new_branch(info, {
> 
>                 path_style = function()
> 
>                     return vim.fs.basename(info.new_branch)
> 
>                 end,
> 
>                 branch_style = "prompt",
> 
>             })
> 
>         end}}
> 
> ```
> 
> 
> 
> Welp... Sorry to bother you for nothing then.

Yeah same problem as path, if you are creating a new branch, I can't guess the name of the branch.

But if there's a particular workflow you have in mind, I would love to hear it. Maybe there is a feature I can add which would make it achievable for you.

Like for example, when you create a new branch & worktree from main, what would be the naming formats you are trying to achieve.

**Comment by vunhatchuong** [comment]
In my ideal workflow, the branch name is the priority and the path will inherit from it.

1. `:Arbor add`
2. Type branch name : `fix/nvim-error`.
3. Plugin will auto create folder based on basename. So branch `fix/nvim-error` will create a folder name `nvim-error`.
4. If that folder already exists then prompt for user input.

Also I think there should be rollback functionality. Currently, during `:Arbor add` if you create a new branch but the folder already exists, the branch is still created so now I have to manually fix it or create the worktree myself.

**Comment by xavierchanth** [comment]
> In my ideal workflow, the branch name is the priority and the path will inherit from it.
> 
> 
> 
> 1. `:Arbor add`
> 
> 2. Type branch name : `fix/nvim-error`.
> 
> 3. Plugin will auto create folder based on basename. So branch `fix/nvim-error` will create a folder name `nvim-error`.
> 
> 4. If that folder already exists then prompt for user input.
> 
> 
> 
> Also I think there should be rollback functionality. Currently, during `:Arbor add` if you create a new branch but the folder already exists, the branch is still created so now I have to manually fix it or create the worktree myself.

Ah, makes sense. I'll need some time to think about edge cases, but I should be able to make some time over the weekend to come up with a good solution for you.

**Comment by xavierchanth** [comment]
@vunhatchuong please check out https://github.com/XavierChanth/arbor.nvim/releases/tag/0.3.0

**Comment by vunhatchuong** [comment]
Thanks! I'm able to make my workflow work with

```lua
require("arbor").actions.add_new_branch(info, {
    path_style = function()
        return vim.fs.basename(info.new_branch)
    end,
    branch_style = "prompt",
})
```

The new `path_style="branch"` doesn't work for me since it creates the new folder nested, so branch with the name `fix/nvim-error` will create a new folder with the path`./fix/nvim-error` instead of `./nvim-error`

There's still the problem with rollback tho.

**Comment by xavierchanth** [comment]
> There's still the problem with rollback tho.

I'm not sure what you mean by this.

**Comment by vunhatchuong** [comment]
I was referring to this part of my comment:

> Also I think there should be rollback functionality. Currently, during :Arbor add if you create a new branch but the folder already exists, the branch is still created so now I have to manually fix it or create the worktree myself.

**Comment by xavierchanth** [comment]
I'm hesitant to implement anything that deletes branches. I don't want this plugin responsible deleting your work by mistake.

I considered having an option for remove worktree to also remove branches, but I erred on the side of caution.

If I have a good testing solution then I'll consider it, but the test suite is a big time sink. Since some flows require user interaction, and the test harness needs create and test against   repositories of various formats.

**Comment by vunhatchuong** [comment]
> I'm hesitant to implement anything that deletes branches

Maybe you can add a prompt like "Failed: folder already exists, do you want to delete the {info.new_branch}?"

**Comment by xavierchanth** [comment]
> > I'm hesitant to implement anything that deletes branches
> 
> Maybe you can add a prompt like "Failed: folder already exists, do you want to delete the {info.new_branch}?"

@vunhatchuong I've added a hook called `on_add_failed` to the add options.
arbor exposes a simple wrapper around plenary which will resolve the git binary that you've configured it with:

```lua
local function delete_branch(info, branch)
  local job = require("arbor").git.job({
    args = {"branch", "-d", branch},
   })
   job:start()
end
```

**I did not test this hook snippet**

You can build your own action from here, use the snippet at your own risk.
